docs/source/en/guides/upload.md

```diff
@@ -103,6 +103,80 @@ set, files are uploaded at the root of the repo.
+
+If you want more technical details about how `upload_large_folder` is implemented under the hood, please have a look to the [`upload_large_folder`] package reference.
+
+Here is how to use [`upload_large_folder`] in a script. The method signature is very similar to [`upload_folder`]:
+
+```py
+>>> api.upload_large_folder(
+...     repo_id="HuggingFaceM4/Docmatix",
+...     repo_type="dataset",
+...     folder_path="/path/to/local/docmatix",
+... )
+```
+
+You will see the following output in your terminal:
+```
+Repo created: https://huggingface.co/datasets/HuggingFaceM4/Docmatix
+Found 5 candidate files to upload
+Recovering from metadata files: 100%|█████████████████████████████████████| 5/5 [00:00<00:00, 542.66it/s]
+
+---------- 2024-07-22 17:23:17 (0:00:00) ----------
+Files:   hashed 5/5 (5.0G/5.0G) | pre-uploaded: 0/5 (0.0/5.0G) | committed: 0/5 (0.0/5.0G) | ignored: 0
+Workers: hashing: 0 | get upload mode: 0 | pre-uploading: 5 | committing: 0 | waiting: 11
+---------------------------------------------------
+```
+
+First, the repo is created if it didn't exist before. Then, the local folder is scanned for files to upload. For each file, we try to recover metadata information (from a previously interrupted upload). From there, it is able to launch workers and print an update status every 1 minute. Here, we can see that 5 files have already been hashed but not pre-uploaded. 5 workers are pre-uploading files while the 11 others are waiting for a task.
+
+A command line is also provided. You can define the number of workers and the level of verbosity in the terminal:
+
+```sh
+huggingface-cli upload-large-folder HuggingFaceM4/Docmatix --repo-type=dataset /path/to/local/docmatix --num-workers=16
+```
+
+<Tip>
+
+For large uploads, you have to set `repo_type="model"` or `--repo-type=model` explicitly. Usually, this information is implicit in all other `HfApi` methods. This is to avoid having data uploaded to a repository with a wrong type. If that's the case, you'll have to re-upload everything.
+
+</Tip>
+
+<Tip warning={true}>
+
+While being much more robust to upload large folders, `upload_large_folder` is more limited than [`upload_folder`] feature-wise. In practice:
+- you cannot set a custom `path_in_repo`. If you want to upload to a subfolder, you need to set the proper structure locally.
+- you cannot set a custom `commit_message` and `commit_description` since multiple commits are created.
+- you cannot delete from the repo while uploading. Please make a separate commit first.
+- you cannot create a PR directly. Please create a PR first and then commit to it by passing `revision`.
```

Comment:
Why?

Reply:
Because we would have to ensure some logic to ensure the script is resumable. If `create_pr=True` is passed, we would have to use `create_pull_request` and then commit files to it. But if the process is interrupted and the user relaunch it, it would have to automatically resume to the last PR. I didn't want to implement this logic in the first PR as it's quite orthogonal to the upload logic. Since large uploads are meant for long lasting processes, I don't think that manually creating PRs will be a blocker here.

Worth case, the user can create a script to call `create_pull_request` and then `large_upload`. I have added a comment in the docs for that (see https://github.com/huggingface/huggingface_hub/pull/2254/commits/ba6e4a87ef6066d8d9028bcecaf0490d91337ce3).